js/ui/iconGridLayout.js

```diff
@@ -234,8 +234,10 @@ const IconGridLayout = new Lang.Class({
 
         let filename = info.get_filename();
         let userDir = GLib.get_user_data_dir();
-        if (filename && userDir && GLib.str_has_prefix(filename, userDir)) {
-            // only delete .desktop files in the user's local data folder
+        if (filename && userDir && GLib.str_has_prefix(filename, userDir) &&
+            info.get_string('X-Endless-CreatedBy') === 'eos-desktop') {
```

Comment:
You could make the `X-Endless-CreatedBy` a public define in `shell-app.h` and use it from here and `shell-app.c` in the previous commit.